KAFKA-9184: Redundant task creation and periodic rebalances after zombie Connect worker rejoins the group

kkonstantine opened the pull request:
Zombie workers, defined as workers that lose connectivity with the Kafka broker coordinator and get kicked out of the group but don't experience a jvm restart, have been keeping their tasks running. This side-effect is more disrupting with the new Incremental Cooperative rebalance protocol. When such workers return: 
a) they join the group with existing assignment and this leads to redundant tasks running in the Connect cluster, and
b) they interfere with the computation of lost tasks, which before this fix would lead to the scheduled rebalance delay not being reset correctly back to 0. This results in periodic rebalances. 

This fix focuses on resolving the above side-effects as follows: 
* Each worker now tracks its connectivity with the broker coordinator in an unblocking manner. This allows the worker to detect that the broker coordinator is unreachable. The timeout is set to be equal to the heartbeat interval. If during this time the connection remains inactive, the worker will proactively stop all its connectors and tasks and will keep attempting to connect to the coordinator. 
* The incremental cooperative assignor will keep the delay to a positive value as long as it can detect lost tasks. If the set of tasks that are computed as lost becomes empty, the delay will be set to zero and no additional rebalancing will be scheduled. 

Besides the test included in this PR, the improvements are being tested with a framework that deploys a Connect cluster on docker images and introduces network partitions between all or selected workers and the Kafka brokers. 

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

kkonstantine committed:
KAFKA-9184: Add a lightweight method to confirm connection to the broker coordinator is up from WorkerGroupMember

kkonstantine committed:
KAFKA-9184: Check connectivity with broker coordinator in intervals and stop tasks if coordinator is unreachable

kkonstantine committed:
KAFKA-9184: Reset rebalance delay when there are no lost tasks

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/WorkerGroupMember.java:
```diff
@@ -151,10 +151,23 @@ public void stop() {
+     * coordinator.
+     * @return true if there is no active connection to a broker coordinator
+     */
+    public boolean coordinatorUnknown() {
```
Where is this function used? It appears that there was already a `coordinatorUnknown()` function available in the WorkerCoordinator.

kkonstantine replied:
Good catch. Remainder from an earlier attempt, when I thought I needed to expose this to the `DistributedHerder`. Indeed unused.

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinator.java:
```diff
@@ -40,6 +40,7 @@
 import java.util.List;
 import java.util.Map;
 import java.util.Objects;
+import java.util.concurrent.TimeUnit;
```
Is this used?

kkonstantine replied:
Ran gradle with lots of flags off. Fixed now.

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinator.java:
```diff
@@ -124,7 +127,22 @@ public void poll(long timeout) {
+                    log.debug("Broker coordinator is ready");
+                } else {
+                    log.debug("Can not connect to broker coordinator");
+                    if (assignmentSnapshot != null && !assignmentSnapshot.failed()) {
```
What is the significance of this check?
If a worker has one failed task and one running task, will the snapshot be reset?
Will the running task be stopped?

kkonstantine replied:
This is a standard check we have when we check the `assignmentSnapshot`. I prefer to err on the safe side and use the second condition too. 

The snapshot refers to your complete assignment (sets of connectors, tasks, revoked connectors and revoked tasks). The failure here corresponds to whether the assignment was successful or not as a whole. Currently the options are `NO_ERROR` and `CONFIG_MISMATCH`, in which case as the comments explain in the code the worker needs to read to the end of the config log and rejoin. A failed assignment does not have assigned tasks (the sets are empty). Even if there's an issue with assignment overwrites elsewhere, what we solve here remains unaffected.

gharris1727 replied:
Thanks for explaining the meaning of a failed snapshot, I wasn't clear on that before. You explanation sounds reasonable.

gharris1727 reviewed (commented):
Great log messages.
Two nits, and one logic question.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinator.java:
```diff
@@ -124,7 +127,22 @@ public void poll(long timeout) {
+                        assignmentSnapshot.connectors().clear();
+                        assignmentSnapshot.tasks().clear();
+                        assignmentSnapshot.revokedConnectors().clear();
+                        assignmentSnapshot.revokedTasks().clear();
```
Will all of these methods to clear state within the assignment snapshot appear to happen atomically w/r/t when the assignment snapshot is read elsewhere in this class?

kkonstantine replied:
Good point. We don't lock access to this collection. Setting to `null` to avoid any concurrent access issues (from the threads that export metrics potentially).

rhauch reviewed (commented):
Thanks, @kkonstantine. One question below, and is there any way to easily test this from within `IncrementalCooperativeAssignorTest` or `RebalanceSourceConnectorsIntegrationTest`? 

We'd probably have to enable `EmbeddedConnectCluster` to shut down the Kafka cluster for a period of time, perhaps by stopping the Kafka cluster and then bring it back up using the same ports (which might actually be problematic if other tests are run in parallel and their Kafka clusters happen to reuse the same ports). Not sure if this is really feasible. Thoughts?

kkonstantine committed:
KAFKA-9184: Remove unused import

kkonstantine committed:
KAFKA-9184: Set assignmentSnapshot to null after stopping tasks

kkonstantine committed:
KAFKA-9184: Extend embedded kafka cluster to allow restarts and add integration test

kkonstantine committed:
KAFKA-9184: Fix checkstyle

kkonstantine committed:
KAFKA-9184: Remove unused method

kkonstantine committed:
KAFKA-9184: Adapt unit test to the additional debug calls

kkonstantine committed:
KAFKA-9184: Add more specific logs to DistributedHerder

kkonstantine commented:
Thanks for the reviews @rhauch @gharris1727 !
I believe I've addressed all the comments. 
The latest build succeeded on JDK8/Scala 2.11 and JDK 11/Scala 2.12
On JDK 11/Scala 2.13 it failed on a single, unrelated test:
`kafka.admin.ReassignPartitionsClusterTest shouldTriggerReassignmentWithZnodePrecedenceOnControllerStartup FAILED`

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinator.java:
```diff
@@ -124,7 +126,19 @@ public void poll(long timeout) {
+                        log.info("Broker coordinator was unreachable for {}ms. Revoking previous assignment {} to " +
+                                "avoid running tasks while not being a member the group", coordinatorDiscoveryTimeoutMs, assignmentSnapshot);
+                        listener.onRevoked(assignmentSnapshot.leader(), assignmentSnapshot.connectors(), assignmentSnapshot.tasks());
+                        assignmentSnapshot = null;
```
Thanks for improving the atomicity of clearing out the snapshot. However, IIUC the way this class is access by multiple threads means that the various places where `assignmentSnapshot` is read, the read may not be consistent and might result in a NPE. This line is the only place where `assignmentSnapshot` can be assigned to null after being not null, but that means other places where `assignmentSnapshot` is read could potentially produce a NPE. For example, `onJoinComplete(...)` (lines 197-199) does the following:
```
            if (assignmentSnapshot != null) {
                assignmentSnapshot.connectors().removeAll(newAssignment.revokedConnectors());
                assignmentSnapshot.tasks().removeAll(newAssignment.revokedTasks());
```
but if the broker coordinator is lost after the first line but before the second or third lines, the latter would produce a NPE. I think there are a few such places, and it'd be good to make the `assignmentSnapshot` member volatile (to ensure proper read semantics for multiple threads) and to use local variables to ensure consistent reads of the snapshot.

kkonstantine replied:
Valid points. To be on the safe side I've made `assignmentSnapshot` volatile, and wherever it's used I'm using a local copy of the reference instead (`final ExtendedAssignment localAssignmentSnapshot = assignmentSnapshot`)

Testing with faults while this PR builds.

gharris1727 commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectWorkerIntegrationTest.java:
```diff
@@ -180,6 +187,56 @@ public void testRestartFailedTask() throws Exception {
+        connect.kafka().startOnlyKafkaOnSamePorts();
+
+        // Allow for the workers to discover that the coordinator is unavailable
+        Thread.sleep(TimeUnit.SECONDS.toMillis(10));
```
Shouldn't the workers discover that the coordinator is unavailable while it is down?
I'm imagining this test going like this:

1. steady-state workers are running
2. brokers stop
3. workers discover the coordinator is unavailable 
4. workers stop their tasks
5. brokers start
6. workers discover the next coordinator 
7. workers start their tasks
8. workers are running unaffected

kkonstantine replied:
That was a bit misplaced, because I actually need 3 explicit delays (due to current lack of appropriate handles from the kafka and connect embedded clusters). 
1. Bring kafka down, allow workers to discover it's down (heartbeat * 2 + 4 sec)
2. Allow for Kafka to come back up
3. Allow for worker cluster to stabilize after the very last rebalance (delay = 5sec) 

Added another commit.

kkonstantine committed:
KAFKA-9184: Declare assignmentSnapshot volatile and use local copies of the reference

kkonstantine committed:
KAFKA-9184: Add some more time between phases in the integration test

rhauch commented:
Two of the 3 jobs passed with the latest commits, and the other is still running:
* https://builds.apache.org/job/kafka-pr-jdk8-scala2.11/26963/
* https://builds.apache.org/job/kafka-pr-jdk11-scala2.13/3727/

Waiting for the 3rd job to complete (almost done):
* https://builds.apache.org/job/kafka-pr-jdk11-scala2.12/9727/

(This, despite the fact that Jenkins previously showed the status of all of these jobs as failed, but apparently they were still running. Evidently others have seen similar behavior recently.)

rhauch commented:
Now, all 3 jobs passed!

kkonstantine commented:
Thanks @rhauch for keep track of the build progress. Looks green to me too 😄 
Just a note that this fix applies to 2.3, 2.4 and trunk release branches. 

If we'd like to merge the zombie fix further back, it's better to issue a separate PR. 
Cheers!

rhauch reviewed (approved):
LGTM. Thanks for all the hard work on this PR, @kkonstantine. And thanks to @gharris1727 and @ncliang for help with testing to identify this.

rhauch commented:
Run with Connect system tests all passed: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/3614/

gharris1727 reviewed (approved):
LGTM.

Thanks @kkonstantine !
